ui/app/services/permissions.js

```diff
@@ -108,33 +108,60 @@ export default Service.extend({
     }
   },
 
-  hasPermission(pathName) {
+  hasPermission(pathName, capabilities = []) {
     const path = this.pathNameWithNamespace(pathName);
 
-    if (this.canViewAll || this.hasMatchingExactPath(path) || this.hasMatchingGlobPath(path)) {
+    if (this.canViewAll) {
       return true;
     }
-    return false;
+
+    if (capabilities.length) {
```

Comment:
At a micro level (not sure if you meant this particular line or the entire function body), another option would be to not specify a capabilities _default_ argument, so:

```javascript
hasPermission(pathName, capabilities) {
```

and then do a type check on the variable below instead:

```javascript
if(typeof capabilities !== 'undefined') {
```

I actually prefer what you've done here though (at a micro level). Adding the default arguments gives a 'clue' as to what you should be passing as an argument, essentially a type hint. Immediately from looking at this code I know the second argument passed to this function should be an array. If you wanted to be super correct maybe add a greater than zero?

```javascript
if (capabilities.length > 0) {
```

In fact, `every` is probably a noop if the array is zero length, but you need to make sure the rest of the body of the function takes all this into account.

I think maybe overall the entire function body (at a macro level) could possibly be a little different if you wanted to, there's quite a bit of conditional logic in here that you could reduce down if you wanted to do that. It's one of those though, I'm not sure if it would make it any clearer what's happening here, it might though. 

```javascript
hasPermission(pathName, capabilities = [null]) {
  const path = this.pathNameWithNamespace(pathName);
  return capabilities.every(capability => this.hasMatchingExactPath(path, capability) || 
  this.hasMatchingGlobPath(path, capability))
}
```

Something like that ^ could be quite fun? Up to you really. Not sure if that works like..

Anyway, this ended up being a little longer than I thought it was gonna be! 😂 Will try and do a proper review of the rest of this today at somepoint. I think there's another few PRs I could hit also so might do a 'Review Friday'

Reply:
hey, thanks john! this is exactly the feedback i was looking for. i was hoping to reduce the amount of conditional logic in the function body, and i like the solution you proposed with setting capabilities as a `null` array by default. thank you!